ql/pricingengines/forward/mcforwardvanillaengine.hpp

```diff
@@ -0,0 +1,171 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+/*! \file mcforwardvanillaengine.hpp
+    \brief Monte Carlo engine for forward-starting strike-reset vanilla options
+*/
+
+#ifndef quantlib_mcforwardvanilla_engine_hpp
+#define quantlib_mcforwardvanilla_engine_hpp
+
+#include <ql/pricingengines/mcsimulation.hpp>
+#include <ql/instruments/forwardvanillaoption.hpp>
+#include <ql/instruments/vanillaoption.hpp>
+
+namespace QuantLib {
+
+    //! Monte Carlo engine for forward-starting vanilla options
+    /*! \ingroup forwardengines
+    */
+    template<template <class> class MC,
+             class RNG = PseudoRandom, class S = Statistics>
+    class MCForwardVanillaEngine : public GenericEngine<ForwardOptionArguments<VanillaOption::arguments>,
+                                                        VanillaOption::results>,
+                                   public McSimulation<MC,RNG,S>
+    {
+      public:
+        typedef typename McSimulation<MC,RNG,S>::path_generator_type
+            path_generator_type;
+        typedef typename McSimulation<MC,RNG,S>::path_pricer_type
+            path_pricer_type;
+        typedef typename McSimulation<MC,RNG,S>::stats_type
+            stats_type;
+        // constructor
+        MCForwardVanillaEngine(
+             const ext::shared_ptr<StochasticProcess>& process,
+             Size timeSteps,
+             Size timeStepsPerYear,
+             bool brownianBridge,
+             bool antitheticVariate,
+             bool controlVariate,
```

Comment:
The machinery for control variates is not used by the derived engines, so we might strip it off (in my old age, I start to see it as "untested code" instead of "code that might be useful some day"&mdash;it usually isn't).  I'm not dead set against it, though.  What do you think?

Reply:
That sounds good to me.

For forward start options where the reset date is much closer than the expiry date, the vanilla expiring on the same date might be a reasonably good control variate - but it would certainly need extra testing, something for another commit later perhaps!

Also probably made unnecessary if the semi-analytical pricer becomes available for these options.